lib/pure/collections/sets.nim

```diff
@@ -571,10 +571,25 @@ proc map*[A, B](data: HashSet[A], op: proc (x: A): B {.closure.}): HashSet[B] =
   result = initHashSet[B]()
   for item in items(data): result.incl(op(item))
 
+from std/algorithm import sort
+
 proc hash*[A](s: HashSet[A]): Hash =
   ## Hashing of HashSet.
-  for h in 0 .. high(s.data):
-    result = result xor s.data[h].hcode
+  # This handles tombstones (iterating over all `s.data` would be wrong).
+  # Iterating over items(s) requires a commutative hash combiner like `xor`
+  # to avoid depending on order (which could differ for 2 HashSet's with different
+  # `data.len` but same elements, after insertions and deletions). But `xor`
+  # has bad mixing properties (eg, would give 0 if HashSet contains
+  # a, b such that hash(a) == hash(b) and a != b). `sort` should have low
+  # overhead compared to the surrounding code.
+  when true:
+    var s2: seq[Hash]
+    for h in s: s2.add hash(h)
+    s2.sort
+    for h in s2: result = result !& hash(h)
+  else:
+    for h in s:
```

Comment:
dead code here

Reply:
it was intentional for clariry, but fine, I've changed it to commented out code.